### PR TITLE
chore: Adding pip-compile-multi et al.

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

Somewhat of a proof-of-concept this PR leverages `pip-compile-multi` to define the various Python dependencies in an exhaustive and pinned manner (not unlike `pip-tools`; `pip-compile-multi` runs `pip-tools` under the covers) for various environments. Per the updated documentation upgrading Python packages should be fairly trivial as one merely needs to run the following:

```
pip-compile-multi
pre-commit autoupdate --freeze
```

Additionally since there is some overlap between `tox` and `pre-commit` this PR moves `mypy` et al. to `pre-commit` and adds a `pre-commit` environment to `tox` which replaces `isort`, `mypy`, etc. adhering to the DRY principle. Unlike the Git hook the `tox` environment checks all files in the repo (and not just those that have changed) and thus additional configuration was added to exclude certain modules (Alembic migrations) or duplicate names (`superset/__init__/.py` and `superset/bin/superset`)

Finally this PR picks up a number of changes as it seems like `isort` and `mypy` versions may have differed between `pre-commit`, `tox` and/or the Git workflows. Note it seems prudent that we try to leverage `tox` from within the GitHub workflows to ensure consistency between local and remote CI, however in the interim I've added the `pre-commit` check to the GitHub workflows. 

### TEST PLAN

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -15,28 +15,34 @@
 # limitations under the License.
 #
 repos:
--   repo: https://github.com/ambv/black
```
Comment (john-bodley) [inline .pre-commit-config.yaml:18]: ABC.

File: scripts/permissions_cleanup.py [hunk 2]
```diff
@@ -16,13 +16,15 @@
 # under the License.
 from collections import defaultdict
 
-from superset import sm
+from superset import security_manager
```
Comment (john-bodley) [inline scripts/permissions_cleanup.py:19]: This file was not previously type checked.

File: setup.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def get_git_sha():
     packages=find_packages(),
     include_package_data=True,
     zip_safe=False,
-    scripts=["superset/bin/superset"],
+    entry_points={"console_scripts": ["superset=superset.cli:superset"]},
```
Comment (john-bodley) [inline setup.py:68]: `mypy` via `pre-commit` had a conflict between `superset/__init__.py` and this script and thus I opted for an alternative formulation of the script definiton.

File: .pylintrc [hunk 4]
```diff
@@ -25,7 +25,7 @@
 
 # Add files or directories to the blacklist. They should be base names, not
 # paths.
-ignore=CVS
+ignore=CVS,migrations
```
Comment (john-bodley) [inline .pylintrc:28]: Required since `migrations` is now a module so it can be ignored my `mypy`.

**Review by craig-rueda — APPROVED** [review]

**Comment by john-bodley** [comment]
@ktmud I'm running into CI issues with Fossa and I speculate it may be reading from an old cache. Do you know how I can resolve these issues?

**Comment by ktmud** [comment]
> @ktmud I'm running into CI issues with Fossa and I speculate it may be reading from an old cache. Do you know how I can resolve these issues?

I think it's configured here: https://github.com/apache/incubator-superset/blob/291306392443a5a0d0e2ee0cc4a95d37c56d4589/.fossa.yml#L34

**Comment by ktmud** [comment]
> @ktmud I'm running into CI issues with Fossa and I speculate it may be reading from an old cache. Do you know how I can resolve these issues?

Just pushed a [fix](https://github.com/apache/incubator-superset/pull/10499/commits/dfbb741cad9d15e8a87586f4d45fe7f56028179c) to your branch. Hope it is what you wanted!

**Review by bkyryliuk — APPROVED** [review]
that's a nice improvement, it can also help with the optional dependencies, e.g. we can have base superset setup & full one with all the extras.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -15,29 +15,35 @@
...
+    hooks:
+      - id: check-added-large-files
+      - id: check-docstring-first
+      - id: check-yaml
         exclude: ^helm/superset/templates/
-    -   id: debug-statements
+      - id: debug-statements
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+        args: ["--markdown-linebreak-ext=md"]
```
Comment (john-bodley) [inline .pre-commit-config.yaml:44]: I've been tearing my hair out with the `trailing-whitespace` hook. The hook works locally but complains in a remote setting stating that `README.md` needs reformatting albeit the fact that by construction the environments/package versions are equivalent. The only way I was able to resolve this was with the  `--markdown-linebreak-ext` argument.

**Comment by hajdbo** [comment]
now that superset/bin/superset has been replaced, how do we "bin/superset db upgrade" ?

**Comment by john-bodley** [comment]
@hajdbo if you've installed Superset, you should run `superset db upgrade`.

**Comment by hajdbo** [comment]
> @hajdbo if you've installed Superset, you should run `superset db upgrade`.

I used to run it from the repo, not docker or pip. What/where is "superset" now?

**Comment by hajdbo** [comment]
> > @hajdbo if you've installed Superset, you should run `superset db upgrade`.
> 
> I used to run it from the repo, not docker or pip. What/where is "superset" now?

found it in the pipenv directory, all good.

File: requirements/base.txt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+linear-tsv==1.1.0         # via tabulator
+mako==1.1.3               # via alembic
+markdown==3.2.2           # via apache-superset
+markupsafe==1.1.1         # via jinja2, mako, wtforms
+marshmallow-enum==1.5.1   # via flask-appbuilder
+marshmallow-sqlalchemy==0.23.1  # via flask-appbuilder
+marshmallow==3.7.1        # via flask-appbuilder, marshmallow-enum, marshmallow-sqlalchemy
+msgpack==1.0.0            # via apache-superset
+multidict==4.7.6          # via aiohttp, yarl
+mysqlclient==1.4.2.post1  # via apache-superset
```
Comment (robdiciuccio) [inline requirements/base.txt:79]: `mysqlclient` is GPL licensed, and can't be included in the base requirements. It seems the new file structure is not working with FOSSA license checks, which we also need to resolve. cc @villebro @craig-rueda @dpgaspar for the 0.38 release

File: requirements/base.txt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+numpy==1.19.1             # via pandas, pyarrow
+openpyxl==3.0.4           # via tabulator
+packaging==20.4           # via bleach
+pandas==1.0.5             # via apache-superset
+parsedatetime==2.6        # via apache-superset
+pathlib2==2.3.5           # via apache-superset
+pillow==7.2.0             # via apache-superset
+polyline==1.4.0           # via apache-superset
+prison==0.1.3             # via flask-appbuilder
+psycopg2-binary==2.8.5    # via apache-superset
```
Comment (robdiciuccio) [inline requirements/base.txt:90]: `psycopg2` is LGPL

File: requirements/base.txt (hunk 6 above)
Comment (john-bodley) [reply requirements/base.txt:79]: @robdiciuccio et al. I wonder if we should switch to using the pure Python [PyMySQL](https://github.com/PyMySQL/PyMySQL) driver (potentially works better with `gevent`) which is under the MIT license.

Comment (john-bodley) [reply requirements/base.txt:79]: Note @robdiciuccio `mysqlclient` and `psycopg2` originate from `setup.py`.

**Pull request merged**